File: src/lib/entry/command-log-entry-factory.ts
import { COMMAND_CLASS, EntryType } from "../globals"
import CommandLogEntry from "./command-log-entry"

export default class CommandLogEntryFactory {
    static fromU8(u8: Uint8Array): CommandLogEntry {
        const entryType: number | undefined = u8.at(0)
        if (entryType !== EntryType.COMMAND) {
            throw new Error(`Invalid entryType: ${entryType}`)
        }

        const commandName: number | undefined = u8.at(1)

        if (commandName === undefined || !(commandName in COMMAND_CLASS)) {
            throw new Error(`Invalid commandName: ${commandName}`)
        } else {
            return new COMMAND_CLASS[commandName]({
                commandNameU8: new Uint8Array(u8.buffer, u8.byteOffset + 1, 1),
                commandValueU8: new Uint8Array(u8.buffer, u8.byteOffset + 2, u8.byteLength - 2),
            })
        }
    }
}

File: external/logsrdcpp/src/entry/command/CommandLogEntryFactory.cpp
#include "CommandLogEntryFactory.h"
#include "CreateLogCommand.h"
#include "SetConfigCommand.h"
#include "../../Globals.h"
#include <stdexcept>

namespace logsrd {

std::unique_ptr<CommandLogEntry> CommandLogEntryFactory::fromU8(
    std::span<const uint8_t> data) {
    if (data.empty() || data[0] != TYPE_BYTE_COMMAND) {
        throw std::runtime_error("Invalid entryType");
    }
    if (data.size() < 2) {
        throw std::runtime_error("Invalid commandName: undefined");
    }

    uint8_t commandName = data[1];
    auto nameBytes = std::vector<uint8_t>{commandName};
    auto valueBytes = std::vector<uint8_t>(data.begin() + 2, data.end());

    JSONCommandTypeArgs args;
    args.commandNameU8 = std::move(nameBytes);
    args.commandValueU8 = std::move(valueBytes);

    switch (static_cast<CommandName>(commandName)) {
    case CommandName::CREATE_LOG:
        return std::make_unique<CreateLogCommand>(std::move(args));
    case CommandName::SET_CONFIG:
        return std::make_unique<SetConfigCommand>(std::move(args));
    default:
        throw std::runtime_error("Invalid commandName: " + std::to_string(commandName));
    }
}

} // namespace logsrd
